numba_rvsdg/core/datastructures/scfg.py

```diff
@@ -781,23 +783,110 @@ def from_dict(graph_dict: dict):
             Dictionary of block names in YAML string corresponding to their
             representation/unique name IDs in the SCFG.
         """
-        scfg_graph = {}
         name_gen = NameGenerator()
-        block_dict = {}
-        for index in graph_dict.keys():
-            block_dict[index] = name_gen.new_block_name(block_names.BASIC)
-        for index, attributes in graph_dict.items():
-            jump_targets = attributes["jt"]
-            backedges = attributes.get("be", ())
-            name = block_dict[index]
-            block = BasicBlock(
-                name=name,
-                backedges=tuple(block_dict[idx] for idx in backedges),
-                _jump_targets=tuple(block_dict[idx] for idx in jump_targets),
-            )
-            scfg_graph[name] = block
-        scfg = SCFG(scfg_graph, name_gen=name_gen)
-        return scfg, block_dict
+        block_ref_dict = {}
+
+        blocks = graph_dict["blocks"]
+        edges = graph_dict["edges"]
+        backedges = graph_dict["backedges"]
+        if backedges is None:
+            backedges = {}
+
+        for key, block in blocks.items():
+            assert block["type"] in block_types
+            block_ref_dict[key] = key
+
+        # Find head of the graph, i.e. node which isn't in anyones contains
+        # and no edges point towards it (backedges are allowed)
+        heads = set(blocks.keys())
+        for block_name, block_data in blocks.items():
+            if block_data.get("contains"):
+                heads.difference_update(block_data["contains"])
+            jump_targets = set(edges[block_name])
+            if backedges.get(block_name):
+                jump_targets.difference_update(set(backedges[block_name]))
+            heads.difference_update(jump_targets)
+        assert len(heads) > 0
+
+        seen = set()
+
+        def make_scfg(curr_heads: set, exiting: str = None):
```

Comment:
I think this should not be a closure, unless it really needs to be.

Reply:
I managed to write this such that it isn't a closure.

Reply:
I investigated this a bit more and I think this function will need some significant changes so as to make it more readable and that it becomes clearer what goes on. I think the `make_scfg` shouldn't be a closure and I also think that some stuff can be decomposed into smaller functions too.